packages/core/src/tracing/instrumentation/cloud/aws/s3.js

```diff
@@ -0,0 +1,215 @@
+  /**
+   * Original args for ALL AWS SDK Requets: [method, params, callback]
+   */
+
+  if (!isOperationShimmable(originalArgs[0]) || typeof originalArgs[2] !== 'function') {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  if (cls.tracingSuppressed()) {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  const parentSpan = cls.getCurrentSpan();
+
+  if (!parentSpan || isExitSpan(parentSpan)) {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  return cls.ns.runAndReturn(() => {
+    const span = cls.startSpan('s3', EXIT);
+    span.ts = Date.now();
+    span.stack = tracingUtil.getStackTrace(instrumentedMakeRequest);
+    span.data.s3 = buildS3Data(originalArgs[0], originalArgs[1]);
+
+    const _originalCallback = originalArgs[2];
+
+    originalArgs[2] = cls.ns.bind(function(err, data) {
+      finishSpan(err, data, span);
+      setImmediate(() => finishSpan(err, data, span));
+      return _originalCallback.apply(this, arguments);
+    });
+
+    return originalMakeRequest.apply(ctx, originalArgs);
+  });
+}
+
+function shimRequestPromise(originalPromise) {
+  return function shimmedPromise() {
+    if (isActive && isOperationShimmable(this.operation)) {
+      const originalArgs = new Array(arguments.length);
+      for (let i = 0; i < originalArgs.length; i++) {
+        originalArgs[i] = arguments[i];
+      }
+
+      return instrumentedRequestPromise(this, originalPromise, originalArgs);
+    }
+    return originalPromise.apply(this, arguments);
+  };
+}
+
+function instrumentedRequestPromise(ctx, originalPromise, originalArgs) {
+  if (cls.tracingSuppressed()) {
+    return originalPromise.apply(ctx, originalArgs);
+  }
+
+  if (!isOperationShimmable(ctx.operation)) {
+    return originalPromise.apply(ctx, originalArgs);
+  }
+
+  return cls.ns.runAndReturn(function() {
+    const span = cls.startSpan('s3', EXIT);
+    span.ts = Date.now();
+    span.stack = tracingUtil.getStackTrace(instrumentedRequestPromise);
+    const _promise = originalPromise;
+
+    originalPromise = cls.ns.bind(function patchedPromise() {
+      span.data.s3 = buildS3Data(this.operation, this.params);
+
+      return new Promise((resolve, reject) => {
+        _promise
+          .apply(this, arguments)
+          .then(data => {
+            resolve(data);
+            finishSpan(null, data, span);
+            setImmediate(() => finishSpan(null, data, span));
+          })
+          .catch(err => {
+            reject(err);
+            finishSpan(err, null, span);
+            setImmediate(() => finishSpan(err, null, span));
+          });
+      });
+    });
+
+    return originalPromise.apply(ctx, originalArgs);
+  });
+}
+
+function finishSpan(err, _data, span) {
+  if (err) {
+    addErrorToSpan(err, span);
+  }
+
+  span.d = Date.now() - span.ts;
+  span.transmit();
+}
+
+function addErrorToSpan(err, span) {
+  if (err) {
+    span.ec = 1;
+    if (err.code) {
+      span.data.s3.error = err.code;
```

Comment:
Do other S3 instrumentations capture the code or rather an error message?

Reply:
Not sure if I got the question right. the way it is right now, which seems to be the standard for all AWS lib, err.code gives a string. eg: UnexpectedParameter. I can maybe send a mix of both `code` and `message`, or maybe even better: `err.message || err.code`. Did I get your point right? @basti1302